// limit-queue/src/RateLimitQueue.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
class RateLimitQueue {
    constructor(cap, poolMs) {
        this.cap = cap;
        this.poolMs = poolMs;
        this.items = [];
        this.numDone = 0;
        this.poolPromise = undefined;
    }
    limit(func) {
        this.items.push(func);
        return this.process();
    }
    getItemCount() {
        return this.items.length;
    }
    process() {
        if (!this.poolPromise) {
            this.poolPromise = new Promise((resolve, reject) => {
                setTimeout(() => {
                    this.poolPromise = undefined;
                    this.numDone = 0;
                    resolve();
                }, this.poolMs);
            });
        }
        if (this.numDone >= this.cap) {
            return this.poolPromise.then(() => this.process());
        }
        const nextItem = this.items.shift();
        this.numDone++;
        return nextItem();
    }
}
exports.RateLimitQueue = RateLimitQueue;
